---
title: bioconductor 包的教程还是一如既往的烦
author: Zhuoer Dong
date: '2018-12-16'
slug: bioconductor-vignette-annoy-me-as-usual
categories: 2018
tags: []
authors: []
---

<script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */
</style>


<p>今天下午在看一个 Bioconductor 包的 <a href="https://www.bioconductor.org/packages/release/workflows/vignettes/simpleSingleCell/inst/doc/work-1-reads.html">vignette</a>，做 single cell 分析的。</p>
<p>我发现我现在真是对 <code>read.delim()</code> 之流充满了厌恶。还有 <code>library(SingleCellExperiment)</code> ，要等几秒就不说的，还把我的 R seession 搞得像爆炸了一样。</p>
<p>具体来说，我觉得这个教程中不好的地方是：</p>
<ol style="list-style-type: decimal">
<li><p>真逗，你还不运行了啊，就不会按名称 join 吗？</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="fu">match</span>(<span class="fu">colnames</span>(sce), metadata[[<span class="st">&quot;Source Name&quot;</span>]]) <span class="co"># Enforcing    identical order.</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all</span>(<span class="sc">!</span><span class="fu">is.na</span>(m)))</span></code></pre></div></li>
<li><p>坑，你确定原始的 character 转化为 factor 后的 level 一定是你指定的这样？</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(pheno) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;induced&quot;</span>, <span class="st">&quot;control&quot;</span>)</span></code></pre></div></li>
</ol>
<blockquote>
<p>可重复性不仅仅是别人能跑，还要让代码简洁易懂，要有可扩展性。</p>
</blockquote>
<p>最后吐槽一下，还 simpleSingleCell，光是一个 vignette 就这么臭长了，总共有10篇vignette，谁想学啊？顺便立个flag，希望我写的 thesis 的 4个 R 包的文档，能让读者学的时候能真正感到 <em>simple</em> 。</p>
<hr />
<p>附一下代码（写了几个钟头，就这么删掉太可惜了）：</p>
<pre><code>---
title: &quot;Analyzing single-cell RNA-seq data containing read counts&quot;
output: html_document
---


```{r}
bfc &lt;- BiocFileCache::BiocFileCache(&quot;raw_data&quot;, ask = FALSE)

lun.zip &lt;- BiocFileCache::bfcrpath(
    bfc, 
    file.path(&quot;https://www.ebi.ac.uk/arrayexpress/files&quot;, &quot;E-MTAB-5522/E-MTAB-5522.processed.1.zip&quot;)
)
lun.sdrf &lt;- BiocFileCache::bfcrpath(
    bfc, 
    file.path(&quot;https://www.ebi.ac.uk/arrayexpress/files&quot;, &quot;E-MTAB-5522/E-MTAB-5522.sdrf.txt&quot;)
)

unzip(lun.zip)
```


```{r message=FALSE, warning=FALSE}
plate1 &lt;- readr::read_tsv(&quot;counts_Calero_20160113.tsv&quot;)
plate2 &lt;- readr::read_tsv(&quot;counts_Calero_20160325.tsv&quot;)
plate &lt;- dplyr::inner_join(plate1, plate2)

meta_data &lt;- readr::read_tsv(lun.sdrf) %&gt;% 
    dplyr::select(&#39;cell&#39; = &#39;Source Name&#39;, &#39;plate&#39; = &#39;Factor Value[block]&#39;, 
                  &#39;pheno&#39; =  &#39;Factor Value[phenotype]&#39;) %&gt;%
    dplyr::mutate(plate = factor(plate)) %&gt;%
    dplyr::mutate(induced = stringr::str_detect(pheno, &#39;induced&#39;)) %&gt;%
    dplyr::mutate(oncogene = ifelse(induced, &#39;induced&#39;, &#39;control&#39;)) %&gt;%
    dplyr::mutate(oncogene = factor(oncogene, c(&#39;induced&#39;, &#39;control&#39;)))

gene_length &lt;- plate$Length;
count_df &lt;- plate[, -2] %&gt;% as.data.frame() %&gt;% tibble::column_to_rownames(&#39;GeneID&#39;)
meta_data %&lt;&gt;% dplyr::left_join(tibble::tibble(cell = colnames(count_df)), .) # reorder

sce &lt;- SingleCellExperiment::SingleCellExperiment(
    assays  = list(counts = as.matrix(count_df)),
    rowData = list(GeneLength = plate$Length, ENSEMBL = rownames(plate)),
    colData = list(Plate = meta_data$plate, Oncogene = meta_data$oncogene)
) %T&gt;% print
```



```{r}
# remove another set of spike-in transcripts, the Spike-In RNA Variants (SIRV) set
sce = sce[!stringr::str_detect(rownames(sce), &#39;^SIRV&#39;), ]

SingleCellExperiment::isSpike(sce, &#39;ERCC&#39;) &lt;- stringr::str_detect(rownames(sce), &#39;^ERCC&#39;)
```


```{r message=F}
SummarizedExperiment::rowData(sce)$SYMBOL &lt;- AnnotationDbi::mapIds(
    org.Mm.eg.db::org.Mm.eg.db, keys = rownames(sce), keytype = &quot;ENSEMBL&quot;, column=&quot;SYMBOL&quot;
)
```
</code></pre>
